coderd/database/migrations/000205_unique_linked_id.up.sql

```diff
@@ -0,0 +1,18 @@
+-- Remove the linked_id if two user_links share the same value.
+-- This will affect the user if they attempt to change their settings on
+-- the oauth/oidc provider. However, if two users exist with the same
+-- linked_value, there is no way to determine correctly which user should
+-- be updated. Since the linked_id is empty, this value will be linked
+-- by email.
+UPDATE ONLY user_links out
+SET
+	linked_id =
+		CASE WHEN (
+			  -- When the count of linked_id is greater than 1, set the linked_id to empty
+			  SELECT  COUNT(*)
+			  FROM    user_links inn
+			  WHERE   out.linked_id = inn.linked_id AND out.login_type = inn.login_type
+		  ) > 1 THEN '' ELSE out.linked_id END;
+
+-- Enforce unique linked_id constraint on non-empty linked_id
+CREATE UNIQUE INDEX idx_user_link_linked_id ON user_links USING btree (linked_id, login_type) WHERE (linked_id != '');
```

Comment:
Postgres typically names these with `idx` as suffix, this would be named ~ `user_links_linked_id_login_type_idx`.

Also, what's the purpose for allowing this index to be partial, could we not delete the conflicting `user_links` entries instead?

Reply:
Unfortunately no. Legacy accounts do not have a `linked_id`, and at present, we make no attempt to update those values when they log in again.

I debated implementing some code to migrate users to the modern as they log in, but that changes the match behavior. At present, if your linked_id is empty, we use your email to match.

You could fathom a situation where someone has more than 1 github account, and more than 1 coder account. If I attempted to fix this, the github account would get bound to the next account logged in, which is not deterministic.

So I'd rather just leave the legacy accounts as they were and keep them empty.